sdk/lib/rtl/priv.c

```diff
@@ -486,3 +486,78 @@ RtlAdjustPrivilege(IN ULONG Privilege,
 
     return STATUS_SUCCESS;
 }
+
+/*
+ * @implemented
+ */
+NTSTATUS
+NTAPI
+RtlRemovePrivileges(IN HANDLE TokenHandle,
+                    IN PULONG PrivilegesToKeep,
+                    IN ULONG PrivilegeCount)
+{
+    DPRINT("RtlRemovePrivileges(%p, %p, %u)\n", TokenHandle, PrivilegesToKeep, PrivilegeCount);
+
+    /* Save privileges that should be keep */
+    UINT64 PrivilegesToKeepBitmap = 0;
+    if (PrivilegeCount)
+    {
+        ULONG ulCount;
+        PULONG pulPrivilegesToKeep = PrivilegesToKeep;
+        for (ulCount = 0; ulCount < PrivilegeCount; ulCount++)
+        {
+            PrivilegesToKeepBitmap |= (1ULL << *pulPrivilegesToKeep);
+            if (*pulPrivilegesToKeep++ > SE_MAX_WELL_KNOWN_PRIVILEGE)
+            {
+                return STATUS_INVALID_PARAMETER;
+            }
+        }
+    }
+
+    /* Get token privileges information */
+    ULONG ulReturnLength;
+    BYTE tpBuffer[sizeof(TOKEN_PRIVILEGES) +
+                  sizeof(LUID_AND_ATTRIBUTES) * (SE_MAX_WELL_KNOWN_PRIVILEGE - SE_MIN_WELL_KNOWN_PRIVILEGE)];
+    NTSTATUS Status = ZwQueryInformationToken(TokenHandle,
+                                              TokenPrivileges,
+                                              tpBuffer,
+                                              sizeof(tpBuffer),
+                                              &ulReturnLength);
+    if (NT_SUCCESS(Status))
+    {
+
+        /* Remove all privileges that we don't need to keep */
+        PTOKEN_PRIVILEGES tp = (PTOKEN_PRIVILEGES)tpBuffer;
+        DWORD i;
+        for (i = 0; i < tp->PrivilegeCount; i++)
+        {
+            UINT64 ulPrivilege = ((UINT64)tp->Privileges[i].Luid.HighPart << 32) | tp->Privileges[i].Luid.LowPart;
```

Comment:
```suggestion
            LARGE_INTEGER Privilege = (LARGE_INTEGER)tp->Privileges[i].Luid;
```
Then the `ulPrivilege` below can be replaced by `Privilege.QuadPart`.

Reply:
It's seems like `LUID` cannot be cast to `LARGE_INTEGER` directly due to they are structrues, shall I use:

`PLARGE_INTEGER Privilege = (PLARGE_INTEGER)&Privileges->Privileges[i].Luid;` instead OR:

`LARGE_INTEGER Privilege;`
`Privilege.HighPart = Privileges->Privileges[i].Luid.HighPart;`
`Privilege.LowPart = Privileges->Privileges[i].Luid.LowPart;`？

Reply:
`PLARGE_INTEGER Privilege = (PLARGE_INTEGER)&Privileges->Privileges[i].Luid;` may be simpler indeed, and it's close to something I was testing (which was:
```c
LARGE_INTEGER Privilege = *(LARGE_INTEGER*)&Privileges->Privileges[i].Luid;
```
)

Reply:
Okay, I get it

Reply:
Honestly the whole code inside the loop looks fishy to me. Why do we have to toy around with privileges in this way, and especially with LARGE_INTEGERs and 64-bit unsigned integers?  This looks quite unnatural and not so human readable.

Reply:
This implement equals to Windows (as Win11 I currently use)

This is a simple and effective way to save all the privileges specified in `PrivilegesToKeep` input parameter, and then we determine each privilege should be keep or not by testing the corresponding bit. If we do that by enumerating `PrivilegesToKeep` again and again, it's not a good idea I think.

Reply:
Yeah, which is why the whole LowPart/HighPart dance makes no sense. So, assuming the logic is correct and should stay like this, just use LowPart for the bit shift and skip this step if HighPart is set (and/or assert that HighPart is zero). Then we can forget about the "how to convert LUID to LARGE_INTEGER" discussion.

Reply:
I intend to keep the same with Windows, so here `HighPart` is involved. But in fact, `HighPart` is always 0 in this case, privilege values won't greater than 63 yet.